modules/ethereum/src/lib.rs

```diff
@@ -623,6 +642,30 @@ impl<T: Trait> Storage for BridgeStorage<T> {
 	}
 }
 
+/// Verify that transaction is included into given finalized block.
+pub fn verify_transaction_finalized<S: Storage>(
+	storage: &S,
+	block: H256,
+	tx_index: u64,
+	proof: &Vec<RawTransaction>,
+) -> bool {
+	if tx_index >= proof.len() as _ {
+		return false;
+	}
+
+	let header = match storage.header(&block) {
+		Some((header, _)) => header,
+		None => return false,
+	};
+	let (finalized_number, finalized_hash) = storage.finalized_block();
+
+	if header.number > finalized_number || (header.number == finalized_number && block != finalized_hash) {
```

Comment:
I think something like this makes it a little easier to understand.

```rust
let is_finalized = header.number == finalized_number && block == finalized_hash;
if header.number > finalized_number || !is_finalized {
    ...
}
```

Also, what if `header.number` is less than `finalized_number`? How do we verify that the block has been finalized?